pkg/controllers/user/globaldns/user_cluster_globaldns_controller.go

```diff
@@ -56,67 +77,130 @@ func (g *UserGlobalDNSController) sync(key string, obj *v3.GlobalDNS) (runtime.O
 	}
-	return nil, err
+
+	//gather endpoints
+	return g.fetchGlobalDNSEndpointsForIngresses(allIngresses, obj)
 }
 
-func (g *UserGlobalDNSController) updateGlobalDNSEndpointsForTarget(t *v3.Target, obj *v3.GlobalDNS) (*v3.GlobalDNS, error) {
-	// each target will have appName, this appName is also the namespace in which all workloads for this app are created
-	ingresses, err := g.ingressLister.List(t.AppName, labels.NewSelector())
-	if err != nil {
-		return nil, err
+func (g *UserGlobalDNSController) reconcileProjects(obj *v3.GlobalDNS) ([]string, error) {
+	// go through target projects which are part of the current cluster and find all ingresses
+	var allIngresses []*v1beta1.Ingress
+
+	for _, projectNameSet := range obj.Spec.ProjectNames {
+		split := strings.SplitN(projectNameSet, ":", 2)
+		if len(split) != 2 {
+			return nil, fmt.Errorf("UserGlobalDNSController: Error in splitting project Name %v", projectNameSet)
+		}
+		// check if the project in this iteration belongs to the same cluster in current context
+		if split[0] != g.clusterName {
+			logrus.Debugf("UserGlobalDNSController: Continuing since project %v doesn't belong in cluster %v", split[1], g.clusterName)
+			continue
+		}
+		projectID := split[1]
+		//list all namespaces in this project and list all ingresses within each namespace
+		namespaces, err := g.namespaceLister.List("", labels.NewSelector())
+		if err != nil {
+			return nil, fmt.Errorf("UserGlobalDNSController: Error listing cluster namespaces")
+		}
+		var namespacesInProject []*corev1.Namespace
+		for _, namespace := range namespaces {
+			nameSpaceProject := namespace.ObjectMeta.Labels[projectSelectorLabel]
+			if strings.EqualFold(projectID, nameSpaceProject) {
+				namespacesInProject = append(namespacesInProject, namespace)
+			}
+		}
+		for _, namespace := range namespacesInProject {
+			ingresses, err := g.ingressLister.List(namespace.Name, labels.NewSelector())
+			if err != nil {
+				return nil, err
+			}
+			allIngresses = append(allIngresses, ingresses...)
+		}
 	}
+	//gather endpoints
+	return g.fetchGlobalDNSEndpointsForIngresses(allIngresses, obj)
+}
+
+func (g *UserGlobalDNSController) fetchGlobalDNSEndpointsForIngresses(ingresses []*v1beta1.Ingress, obj *v3.GlobalDNS) ([]string, error) {
 	if len(ingresses) == 0 {
 		return nil, nil
 	}
 
+	var allEndpoints []string
+	//gather endpoints from all ingresses
 	for _, ing := range ingresses {
 		if gdns, ok := ing.Annotations[annotationGlobalDNS]; ok {
 			// check if the globalDNS in annotation is same as the FQDN set on the GlobalDNS
 			if gdns != obj.Spec.FQDN {
 				continue
 			}
-			//update endpoints on GlobalDNS status field
+			//gather endpoints from the ingress
 			ingressEndpoints := gatherIngressEndpoints(ing.Status.LoadBalancer.Ingress)
-			updatedGDNS, err := g.updateGlobalDNSEndpoints(obj, ingressEndpoints)
-			if err != nil {
-				return updatedGDNS, err
-			}
+			allEndpoints = append(allEndpoints, ingressEndpoints...)
```

Comment:
would it be better if `allEndpoints` was a map here? this way we'd avoid appending duplicate values and we dedup them anyway later on. 
we also convert to map again for checking if endpoints differ

Reply:
Have changed the logic we are updating clusterEndpoints and dns.Endpoints will just be aggregated from all clusterEndpoints. So no need to dedup anymore.